rrtmgp/kernels-openacc/mo_gas_optics_kernels.F90

```diff
@@ -402,10 +402,81 @@ subroutine gas_optical_depths_minor(ncol,nlay,ngpt,        &
     integer  :: icol, ilay, iflav, igpt, imnr
     integer  :: gptS, gptE
     integer  :: minor_start, minor_loc, extent
+
+    real(wp),    dimension(ncol,nlay,ngpt) :: tau_t
+    real(wp) :: myplay, mytlay, mycol_gas_h2o, mycol_gas_imnr, mycol_gas_0
+    real(wp) :: myfminor(2,2)
```

Comment:
Two things I'm confused about here. First, why are all these local variables helpful/necessary? Second, it looks like this new large block of code is for the case in which all the g-point intervals are the same (`all(minor_limits_gpt(2,:) - minor_limits_gpt(1,:) == gpt_diff`). This is true now but quite unlikely to be true generally. Is this a big performance win on the GPUs?

Reply:
The local variables are needed to remove repeated memory reads inside the innermost (`do imnr = 1, extent`) loop. They help reduce the memory traffic quite a bit.
As for the performance, the routine `gas_optical_depths_minor` is the most expensive in long wave example. Unoptimized, it accounts for 60% of the runtime on Piz Daint P100 and 80% of the runtime on V100. Optimized version decreases the time of this routine by about 10x.

What would be the typical g-point intervals? Is it possible to keep them same, or it would be totally unreasonable? If they are not all the same, but pretty close, this optimization should also work, just slightly changed.

Reply:
I'm a little surprised that the compiler doesn't save the result of the memory read. No? If not then we should implement this change in the alternative form of the loop as well. 

I don't think we can promise to keep all the g-point intervals the same. Does it help to know they always be in the range 1 to roughly 16? How narrow does the range of intervals need to be before the helps?

Reply:
It is difficult to know exactly what the compiler would do. It is likely that it will use the registers, but it may think that the register pressure is already high and rely on cache instead.
Regarding the other variant I don't see many opportunities to store stuff in registers there because of a different data access pattern.

I think it is very helpful to know that the intervals are usually less than 16 (or 20, but something small anyway). Then we can always use the optimized variant, and just use the largest interval as `gpt_diff`. But I'll need to see the other typical cases to tell you more

Reply:
We don't have other variants for the moment. They will come in the form of new data but not for a while (few months). We can assume, for the moment, that the maximum number of g-points per band (`gpt_diff`) will not be larger than 32. 

If the code you wrote will be more efficient than the original for values of `gpt_diff` between 1 and 32 then I'd propose you code for variable number and omit the other branch. I don't mind having different kernels for the CPU and GPU versions but having two versions of the loops with the same kernel-y bits inside seems confusing. 

Would you mind?

Reply:
Seems good to me, I'll make the changes now. For `gpt_diff < 32` the new version should definitely be faster. And later with more data I can come back and see if there is anything else I can do to optimize that routine.